### PR TITLE
Allow the @Mock injection of multiple parameters with the same type

### LINKED ISSUE #6: Store is misused in MockitoExtension
I realize that MockitoExtension should not be considered for production use but the way the ExtensionContext.Store is used isn't a good example of why a Store should be used.  With the existing code, it's possible to return two references to the same object if the same Java class is "injected" into a test method twice.  For instance:

```
@ExtendWith(MockitoExtension.class)
public class MockitoExtensionMisuse {
  
  @Test
  public void testSomethingWithTwoProperties(@InjectMock Properties properties1, @InjectMock Properties properties2) {
    ...
  }

}
```

In the testSomethingWithTwoProperties() method, both properties1 and properties2 will be references to the same instance.

### PR DESCRIPTION
**Pull request by smoyer64**
Resolves #6.

### CODE REVIEW

**Comment by marcphilipp** [comment]
Unfortunately, the parameter name is not always [present](http://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Parameter.html#isNamePresent--).

Alternative proposal: Add a `value` parameter to the `@InjectMock` annotation that is used to identify the mock?

File: junit5-mockito-extension/src/test/java/com/example/mockito/MockitoExtensionInBaseClassTest.java [hunk 1]
```diff
@@ -46,4 +48,11 @@ void secondTestWithInjectedMock(@InjectMock MyType myType) {
 		assertEquals(42, numberGenerator.next());
 	}
 
+	@Test
+	void multipleInjectionsOfSameTypeTest(@InjectMock MyType myType1, @InjectMock MyType myType2) {
+		assertNotNull(myType1);
+		assertNotNull(myType2);
+		assertTrue(myType1 != myType2);
```
Comment (marcphilipp) [inline junit5-mockito-extension/src/test/java/com/example/mockito/MockitoExtensionInBaseClassTest.java]: `assertNotSame'`?

**Comment by smoyer64** [comment]
How about the parameter index?

e.g. java.lang.String:1

**Comment by marcphilipp** [comment]
Well, that won't work for the use case where the same mock gets injected into a `@BeforeEach` and a `@Test` method:

```java
@ExtendWith(MockitoExtension.class)
class MockitoExtensionInBaseClassTest {

	@Mock
	private NumberGenerator numberGenerator;

	@BeforeEach
	void initialize(TestInfo testInfo, @InjectMock MyType myType) {
		when(myType.getName()).thenReturn(testInfo.getDisplayName());
		when(numberGenerator.next()).thenReturn(42);
	}

	@Test
	void test(@InjectMock MyType myType) {
		assertEquals("firstTestWithInjectedMock(MyType)", myType.getName());
		assertEquals(42, numberGenerator.next());
	}
}
```

**Comment by sbrannen** [comment]
Mockito already supports explicit names for mocks: `@org.mockito.Mock#name()`.

Since https://github.com/mockito/mockito/issues/438 has already been resolved, we should therefore upgrade the sample to Mockito 2.x and rely on the standard support for mock names. Furthermore, we would want to implement `Parameter` support analogous to `org.mockito.internal.configuration.FieldAnnotationProcessor<A>` for determining the name of the parameter.

**Comment by smoyer64** [comment]
@marcphilipp 

I think having the myType in initialize() and myType in test() return references to the same instance is confusing so I'd tend to inject those mocks as class fields.

@sbrannen 

I've wrote a bit of test code against Mockito 2.0.69-beta and it works great.  Would you like me to alter this pull request to resolve both #5 and #6.

Since (I think) the JUnit 5 and Mockito teams have agreed that the production Mockito extension should be maintained by the Mockito team, how much value is there in making this sample "bullet-proof"?  As an aside, I think JUnit's tend towards isolationism is probably correct since it's (in essence) the base framework of most testing.  On the other hand, if every project adopted that criteria, collaboration would always require a third-party repository.

**Comment by sbrannen** [comment]
The sample most certainly does not need to be bullet proof; however, if we can improve it easily that would then make it easier to hand over to the Mockito team. 😉 

So with that in mind, yes, feel free to tackle #5 and #6 in the same PR!

File: junit5-mockito-extension/build.gradle [hunk 2]
```diff
@@ -40,7 +40,7 @@ junit5 {
 
 dependencies {
 	// Needed as compile since we're building a JUnit 5 extension here
-	compile 'org.mockito:mockito-core:1.+'
+	compile 'org.mockito:mockito-core:2.0.69-beta'
```
Comment (marcphilipp) [inline junit5-mockito-extension/build.gradle]: Does `2.+` work?

**Comment by smoyer64** [comment]
This would now resolve both #5 and #6.

File: junit5-mockito-extension/build.gradle (hunk 2 above)
Comment (smoyer64) [inline junit5-mockito-extension/build.gradle]: I tried that as both 2+ and 2.+ and it didn't work - I don't know Gradle well but assumed it was due to the -beta label.

**Comment by marcphilipp** [comment]
LGTM!

@sbrannen Do you want to take another look?

**Comment by sbrannen** [comment]
Yeah, I'd like to take a closer look before merging. So I hope to be able to check it out tomorrow.

**Comment by sbrannen** [comment]
@smoyer64, a few tips:

1. Please do not check in tests that break the build in a stand-alone commit. Rather, submit the test and the fix in the same commit.
1. Please keep commit titles (the first line of the commit message) to 50-70 characters.
1. Please squash commits whenever feasible.

Otherwise, the PR looks pretty good! 👍 

I'll polish and push it a little later.

**Comment by sbrannen** [comment]
Merged into `master` in 039d920c7ff9767b4dbc522febb685ede0b5d053.

**Comment by sbrannen** [comment]
Regarding the tips I gave above, you can see them applied in the reworked commits that I merged.

**Comment by smoyer64** [comment]
I should have realized that the comments needed to be changed too ... I found the JUnit 5 Eclipse formatter rules and tried to copy the style used in the rest of the project.  It's funny how every project has nuances that you've got to learn ... and how each project develops a lexicon that outsiders have to learn!

In any case, I'll keep trying ;)  Issues #5 and #6 can also be closed.

**Comment by sbrannen** [comment]
Yes, indeed. Keep at it!

Regarding formatting, we basically follow the same rules here as in the `junit5` repo. We just haven't bothered to set up Spotless for these projects, but we might do that at a later date.

**Pull request closed without merging**